### PR TITLE
Casing for Princeton D TF coils

### LINKED ISSUE #448: New paramaetric shape - casing for the TF coils
We don't have any of the nice TF coils in the parametric reactors which is a shame as they look great.

However I think part of the reason for this is because they need cases to make them look correct.

To make a casing component from a TF coil it would be useful to expose a few more attributes of the coil. Perhaps a attribute called high_point could help build the casing.

Here is a sketch where the new varibles, attributes and geometry is in red, while the exisiting tf magnet geometry is in black

Screenshot from 2020-11-09 11-59-42

Any thoughts?

### PR DESCRIPTION
**Pull request by RemDelaporteMathurin**
## Proposed changes

This PR fixes #448 

## Types of changes

What types of changes does your code introduce to the Paramak?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [ ] New tests

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Pep8 applied
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by Shimwell** [comment]
Just tried the example from the tests and got quite a lot of warnings and a slightly broken shape

The internal part looks good

Screenshot from 2020-11-10 21-35-22

Screenshot from 2020-11-10 21-32-54

**Comment by RemDelaporteMathurin** [comment]
> Just tried the example from the tests and got quite a lot of warnings and a slightly broken shape

Commit  [505146c](https://github.com/ukaea/paramak/pull/454/commits/505146c087d56cf17fc933efe0de2cc455ee4d62) should fix this.


image

**Comment by RemDelaporteMathurin** [comment]
> I do get a nice shape now, thanks for that fix.
> 
> The only other thing I noticed was a few warnings print out. Is this anything to worry about

Some of these warning are :
- divisions by zero in the derivative approximation by numpy (but the function add_thickness can handle inf and -inf
- warnings in ODEint in scipy occurring at princeton D shape level

Related to the discussion we had yesterday with @billingsley-john , these warnings are multiplied cause the method find_points is called several times (like 3 or 4 times). Having the hash value thing would solve this.

**Comment by Shimwell** [comment]
Images are looking better, perhaps these can be used in the docs

Screenshot from 2020-11-11 14-09-04
Screenshot from 2020-11-11 14-08-58
Screenshot from 2020-11-11 14-21-28

**Comment by Shimwell** [comment]
Wondering if the vertical_section_offset argument is having the correct behavior, set it to 1 to get the above diagrams

**Comment by Shimwell** [comment]
Do you think it could make use of the  magnet.azimuth_placement_angle the from the magnet argument.

**Comment by Shimwell** [comment]
Perhaps we need a test for the numbe of faces in the resulting geometry

asset len(tf_coil_casing.areas)==9

```
number_of_toroidal_field_coils = 8
angle_offset = (360 / number_of_toroidal_field_coils) / 2.
tf_coil_thickness = 50
tf_coil_distance = 130

tf_coil_casing_distance = tf_coil_distance + 40
tf_coil_casing_thickness = 20


tf_coils = paramak.ToroidalFieldCoilPrincetonD(
    R1=400,
    R2=1500,  # height
    thickness=tf_coil_thickness,
    distance=tf_coil_distance,
    number_of_coils=number_of_toroidal_field_coils,
    rotation_angle=180,
)

tf_coil_casing = paramak.TFCoilCasing(
    magnet=tf_coils,
    distance=tf_coil_casing_distance,
    inner_offset=tf_coil_casing_thickness,
    outer_offset=tf_coil_casing_thickness,
)
```    

Screenshot from 2020-11-11 16-03-43

**Comment by Shimwell** [comment]
I have updated the make magnet set from the examples so that it uses this new shape and also added some docs to the rst files

**Comment by RemDelaporteMathurin** [comment]
> Perhaps we need a test for the number of faces in the resulting geometry

I've modified the leg points and these holes shouldn't appear anymore. Can you try again ? @Shimwell

**Comment by Shimwell** [comment]
I think it might be more broken now

Screenshot from 2020-11-11 21-00-45

**Comment by RemDelaporteMathurin** [comment]
I've fixed the conflicts, but this new component suffers some serious performance issues cause find_points() is costly and is called many times ... maybe @billingsley-john can have a look 👀

**Comment by RemDelaporteMathurin** [comment]
> I've fixed the conflicts, but this new component suffers some serious performance issues cause find_points() is costly and is called many times ... maybe @billingsley-john can have a look 👀
A points_hash_value has been added which should solve this.
We should maybe wait #485 is merged in before merging this PR in.

**Comment by Shimwell** [comment]
Updated the diagram

Screenshot from 2020-11-13 14-14-05

File: docs/source/paramak.parametric_components.rst [hunk 1]
```diff
@@ -399,6 +399,23 @@ PlasmaBoundaries()
...
+TFCoilCasing()
+^^^^^^^^^^^^^^
+
+|TFCoilCasingallstp| |TFCoilCasingaCutstp| |TFCoilCasingsvg|
+
+.. |TFCoilCasingallstp| image:: https://user-images.githubusercontent.com/8583900/98821523-94943f00-2427-11eb-8047-68f2762c56d7.png
+    :width: 200px
+.. |TFCoilCasingaCutstp| image:: https://user-images.githubusercontent.com/8583900/98821532-96f69900-2427-11eb-99e1-e2461be67511.png
+    :width: 130px
+.. |TFCoilCasingsvg| image:: https://user-images.githubusercontent.com/8583900/98822857-3ff1c380-2429-11eb-80ae-83f2ff875fea.png
```
Comment (Shimwell) [inline docs/source/paramak.parametric_components.rst:411]:
```suggestion
.. |TFCoilCasingsvg| image:: https://user-images.githubusercontent.com/8583900/99081345-904c5b00-25ba-11eb-8a4f-956d4ad6bbc0.png
```

Comment (Shimwell) [reply docs/source/paramak.parametric_components.rst:411]: just changing the image to use the new one

**Comment by Shimwell** [comment]
Hoping to try this on the new docker image. I guess it is just a problem with my local install so the docker image might just give me the confidence needed to merge this

**Comment by Shimwell** [comment]
Merging :-)

**Pull request merged**